diff --git a/index.js b/index.js
index 56c26025074cbcf218e3ec4d810662da93f8024e..0a5b3cc9083d2ef02ecdeec9b2978b7ec1ce5a9d 100644
--- a/index.js
+++ b/index.js
@@ -396,7 +396,7 @@ module.exports = class CachePolicy {
         this._assertRequestHasHeaders(req);
 
         // In all circumstances, a cache MUST NOT ignore the must-revalidate directive
-        if (this._rescc['must-revalidate']) {
+        if (this._rescc['must-revalidate'] || this._securityRequiresRevalidation()) {
             return this._evaluateRequestMissResult(req);
         }
 
@@ -619,19 +619,14 @@ module.exports = class CachePolicy {
      * Returns the maximum age (freshness lifetime) of the response in seconds.
      * @returns {number} The max-age value in seconds.
      */
-    maxAge() {
-        if (!this.storable() || this._rescc['no-cache']) {
-            return 0;
-        }
+    _securityRequiresRevalidation() {
+        return !this.storable() || !!this._rescc['no-cache'] ||
+            (this._isShared && (!!this._rescc['proxy-revalidate'] ||
+                (!!this._resHeaders['set-cookie'] && !this._rescc.public && !this._rescc.immutable)));
+    }
 
-        // Shared responses with cookies are cacheable according to the RFC, but IMHO it'd be unwise to do so by default
-        // so this implementation requires explicit opt-in via public header
-        if (
-            this._isShared &&
-            (this._resHeaders['set-cookie'] &&
-                !this._rescc.public &&
-                !this._rescc.immutable)
-        ) {
+    maxAge() {
+        if (this._securityRequiresRevalidation()) {
             return 0;
         }
 
@@ -640,9 +635,6 @@ module.exports = class CachePolicy {
         }
 
         if (this._isShared) {
-            if (this._rescc['proxy-revalidate']) {
-                return 0;
-            }
             // if a response includes the s-maxage directive, a shared cache recipient MUST ignore the Expires field.
             if (this._rescc['s-maxage']) {
                 return toNumberOrZero(this._rescc['s-maxage']);
